Release candidate orange (rebased)

zmbc opened the pull request:
## Release candidate orange (rebased)

### Description
- *Category*: refactor
- *JIRA issue*: ?

Re-applies the changes from #387, #389, and #393 to main, by applying a new PR #418 to make them compatible.

### Testing
- [x] all tests pass (`pytest --runslow`)

albrja committed:
Mic-4865/Refactor/dataset (#387)

Refactors pseudopeople to create a class that encapsulates dataset schema, the data to be noised, a randomness stream, and the missingness attribute in one place.
- *Category*: Refactor
- *JIRA issue*: [MIC-4865](https://jira.ihme.washington.edu/browse/MIC-4865)

-creates new class to hold dataset schema, a pandas dataframe, a randomness stream, and a missingness bool dataframe
-updates fixtures and tests to reflect changes in the refactor

Testing
All tests pass

albrja committed:
Mic-4865/Refactor/rename dataset (#389)

Mic-4865/Refactor/rename dataset

Renames Dataset to Dataset Schema and DatasetData to Dataset.
- *Category*: Refactor
- *JIRA issue*: [MIC-4865](https://jira.ihme.washington.edu/browse/MIC-4865)

-renames current dataset class to dataset_schema
-renames DatasetData to Dataset
-updates docstrings

Testing
All tests pass

albrja committed:
Mic 4568/lint (#393)

Mic-4568/make-format

Adds Makefile to run formatters
- *Category*: Feature
- *JIRA issue*: [MIC-4568](https://jira.ihme.washington.edu/browse/MIC-4568)

-adds Makefile to include the following commands (which can be run from the terminal within your conda environment) for quality of life improvement for development:
    -`make format` which is running black and isort on the codebase
    -`make integration` which is running `pytest --runslow tests/`  
    -`make unit` which is running `pytest --runslow tests/unit/
`

Testing
Ran make format and saw expected behavior.

zmbc committed:
Refactor: perform column noise in-place (#418)

albrja commented:
@zmbc We do not want to merge this now since we would want to do significant testing with this large of a refactor. But this is the release-candidate branch that we should branch off of for all features for the next release.

albrja committed:
Initial commit for randomstate

albrja committed:
Commit before testing

albrja committed:
Add dataset missingness test

albrja committed:
Update test

albrja committed:
Update comment

albrja committed:
Rebase

albrja committed:
PR comments

albrja committed:
Remove vivarium from setup and test suites

albrja committed:
Update based on new package versions

albrja committed:
Lint

stevebachmeier committed:
prep for mypy (#424)

albrja committed:
Update docstrings for seed/CRN usage

albrja committed:
Make mypy happy

albrja committed:
Fix import error

zmbc committed:
Add reproducible bullet point, remove beta notice (#425)

stevebachmeier committed:
resolve mypy union-attr errors (#428)

stevebachmeier committed:
streamline python version support management (#433)

albrja committed:
Mic-5132/no-any-return (#427)

Mic-5132/no-any-return

Mic-5132/no-any-return
- *Category*: Other
- *JIRA issue*: [MIC-5132](https://jira.ihme.washington.edu/browse/MIC-5132)

-fixes instances of no-any-return issues with mypy

albrja committed:
Mic-5155/return-value (#429)

Mic-5155/return-value

Fix return-value mypy errors
- *Category*: Other
- *JIRA issue*: [MIC-XYZ](https://jira.ihme.washington.edu/browse/MIC-5155)

-fix return value mypy errors

patricktnast committed:
Mic 5154 arg type (#430)

* Pin numpy below 2.0

* fix noise type test

* change float to int

* remove arg-types

* fix layeredconfigtree

* fix state column names

* fix token probability

* fix expected noise

* fix target proportions

* fix generator errors

* fix user filters

* lint

* adjust to only float

* switch to generic types

* fix list and tuples builtin

* lint

* fix error

* add Optional

* revert optional

* try to validate overrides first instead

* make user_filters list instead of None

* make overrides Any type

* fix newlyt broken tests

* fix new errors

* remove return

* Revert "Merge remote-tracking branch 'origin/main' into mic-5154-arg-type"

This reverts commit 2b47eaee7cc74415d98e8e338dfe7d9f6222ab8d, reversing
changes made to b8ecfdf0f228cda8533611b7db139e0b336f5bd6.

* fix merge error

* lint

---------

Co-authored-by: albrja <albrja@uw.edu>

stevebachmeier committed:
fix mypy error [dict-item] (#436)

patricktnast committed:
Fix mypy error: overrides (#435)

* fix overrides with property

* put property above

rmudambi committed:
fix mypy attr-defined errors (#437)

* fix mypy attr-defined errors

* update build workflow to use upstream dependency branches

* add layered config tree as a known third party

* don't install layered_config_tree inside the repository

hussain-jafari committed:
fix assignment error

hussain-jafari committed:
keep default value

hussain-jafari committed:
lint

hussain-jafari committed:
check for assignment errors

albrja committed:
Mic-5187/call-overload (#438)

Mic-5187/call-overload

Fixes one instance of call-overload
- *Category*: Feature
- *JIRA issue*: [MIC-5187](https://jira.ihme.washington.edu/browse/MIC-5187)

-fixes one instance of call overload

hussain-jafari committed:
fix arg-type errors

hussain-jafari committed:
convert var explicitly to float

hussain-jafari committed:
type dataframe

hussain-jafari committed:
ignore assignment

hussain-jafari committed:
don't include optional in user_filters

hussain-jafari committed:
don't add default arg

patricktnast committed:
use np.random.Generator instead of np.Generator (#440)

stevebachmeier committed:
fix [has-type] error (#442)

hussain-jafari committed:
operator error

hussain-jafari committed:
import union

hussain-jafari committed:
test operator

patricktnast committed:
Mypy Errors: No Untyped Call (#441)

* use np.random.Generator instead of np.Generator

* fix all errors

* Revert "use np.random.Generator instead of np.Generator"

This reverts commit 8e41c58d53acb11ab4c31e374bfab2d0c647d2c0.

* add mocker

* add return

* remove ignore

* Update pyproject.toml

* revert attempted change

* add space

hussain-jafari committed:
some updates

hussain-jafari committed:
use list instead of iterable

hussain-jafari committed:
lint test_interface.py

hussain-jafari committed:
disable operator

rmudambi committed:
create DataFilter object (#445)

rmudambi committed:
fix no-untyped-def errors (#446)

albrja committed:
Mic-5186/mypy-var-annotated (#447)

Mic-5186/numpy-var-annotated

rmudambi committed:
use a script to find matching dependency branches (#448)

hussain-jafari committed:
don't ignore list-item

stevebachmeier committed:
Fix import-untyped errors (#449)

* install available stubs; skip scipy
* fix union-attr
* remove old/temp fix for Sphinx issue 8127; pin sphinx-rtd-theme>=0.6 to remove theme error
* fix return-value
* fix var-annotaged
* fix no-any-return

rmudambi committed:
Fix misc mypy errors (#453)

* fix misc mypy errors

* fix mypy errors introduced by incremented pyarrow-stubs version

hussain-jafari committed:
noise configuration (#466)

Category: refactor
Define and use noise configuration object.

Testing
All tests pass.

hussain-jafari committed:
remove node value and nesteddict types (#467)

Category: refactor
JIRA issue: MIC-5359

Remove references to NodeValue and NestedDict types from layered config tree.

Testing
mypy passes.

hussain-jafari committed:
use pipe syntax for typing (#468)

Category: typing
JIRA issue: MIC-5440

Use pipe syntax instead of Optional and/or Union.

Testing
mypy succeeds.

albrja committed:
Albrja/mic-5414/vtu-FuzzyChecker (#471)

Albrja/mic-5414/vtu-FuzzyChecker

Import FuzzyChecker from VTU
- *Category*: Test
- *JIRA issue*: [MIC-5414](https://jira.ihme.washington.edu/browse/MIC-5414)

-import FuzzyChecker from VTU

Testing
All tests pass

albrja committed:
Albrja/mic-5414/vtu-FuzzyChecker (#471)

Albrja/mic-5414/vtu-FuzzyChecker

Import FuzzyChecker from VTU
- *Category*: Test
- *JIRA issue*: [MIC-5414](https://jira.ihme.washington.edu/browse/MIC-5414)

-import FuzzyChecker from VTU

Testing
All tests pass

hussain-jafari committed:
layered config tree typing (#473)

Category: typing
JIRA issue: MIC-5376

Typing updates related to LayeredConfigTree errors.

Testing
All tests pass (including slow tests).

rmudambi committed:
fix np.bool_ type

hussain-jafari committed:
parametrizable data fixtures (#479)

Category: feature
JIRA issue: MIC-5519

Create a dataset fixture which will load data of interest.
Add options for parameters to pytest.
Add tests that command line args are parsed correctly.

Testing
Tests pass, including new tests.

hussain-jafari committed:
create and mark release sample data tests (#480)

Category: test

JIRA issue: MIC-5623
Add three integration tests into release tests.
Mark test for release.

Testing
Tests passed and behaved as expected with and without the --release flag.

hussain-jafari committed:
remove duplicate testing (#481)

Category: test
JIRA issue: MIC-5672

Remove one irrelevant test (test_generate_dataset_with_year).
Replace one duplicative test (test_generate_dataset_from_multiple_shards).
Use FuzzyChecker to check if the amount of noising is the same whether using one or multiple shards.

Testing
Ran pytest with and without --runslow.

hussain-jafari committed:
combine integration and release tests (#482)

Category: test
JIRA issue: MIC-5642

Combine integration and release tests into one folder.
Update some variable naming to be more readable.
Error out if user passes both --runslow and --release.

Testing
Ran pytest with no flags, with --runslow, with --release, and with both.

stevebachmeier committed:
Sbachmei/mic 5669/better handle supported python versions (#490)

hussain-jafari committed:
update setup.py

hussain-jafari committed:
batch error logging

hussain-jafari committed:
define pytest logs directory (#491)

Category: feature
JIRA issue: MIC-5504

Add output-dir option when running pytest.
Create profiling directory using this output directory.
Update test_runner to write logs.

Testing
Ran test_runner.py and checked that the logs directory and profiling directory were created and populated as expected.

hussain-jafari committed:
move test to release folder (#492)

Category: feature
JIRA issue: MIC-5866

Move test_column_dtypes to release suite framework.

Testing
Ran tests on 5 datasets (acs, cps, wic, ssa and census)

hussain-jafari committed:
non default values for testing dropping rows with dummy data (#493)

Category: test
JIRA issue: MIC-5867

Add non-default values when testing omit row and do not respond in unit tests with dummy data.

Testing
Ran tests.

hussain-jafari committed:
include non-1 probabilities in guardian duplication (#494)

Category: test
JIRA issue: MIC-5868

Added test parameters for non-1 probability.
Made data larger and used fuzzy checker.

Testing
Ran tests.

hussain-jafari committed:
split and strengthen drop row tests (#495)

Category: test
JIRA issue: MIC-5516
Split out omit row and do not respond.
Keep tests that proper noising types are defined on default configuration.

Testing
Ran tests. Census failed due to known bug but fuzzy checker caught the expected error. ACS and CPS passed when applying simple but temporary bugfix.

zmbc committed:
Note inexactness of do_not_respond row_probability (#496)

stevebachmeier committed:
change dask default client (#499)

hussain-jafari committed:
Hjafari/feature/mic 5517 guardian test (#498)

Category: test
JIRA issue: MIC-5517
Expand duplicate guardian test to full scale.

Testing
Ran test on census on RI and US data.

hussain-jafari committed:
update do not respond fuzzy checker (#500)

Category: bugfix
JIRA issue: MIC-5893

Use a range instead of single value when fuzzy checking do not respond, expanding by 3% on both sides as described here.

Testing
Ran test on acs, cps, and census sample data and acs and cps for full scale data.

hussain-jafari committed:
full scale test missingness (#501)

Category: feature
JIRA issue: MIC-5515

Add full scale test_dataset_missingness.
Typing.
Add default values to overload functions for generating data with dask.

Testing
Ran tests on acs, cps, and wic for RI and USA.

hussain-jafari committed:
set up dask client test fix and add epic to nightly builds (#502)

Category: bugfix
JIRA issue: hotfix

When we run pytest --runslow, there are tests that run dask and set up a client which persists in the next tests. We were making an assumption that there is no dask client in test_set_up_dask_client_default. This PR shuts down a client if one already exists.
Add full scale testing epic to nightly builds.

Testing
Ran pytest --runslow.

hussain-jafari committed:
unpin numpy and support python 3.12 (#504)

unpin numpy and support python 3.12

hussain-jafari committed:
remove phonetic error from MI column (#507)

remove phonetic error from MI column

zmbc committed:
Remove phonetic error from middle initial column (#505)

stevebachmeier committed:
Sbachmei/mic 6020/fix breaking test (#513)

* fix test_set_up_dask_clean_existing_cluster

* ignore mypy error

* check cluster prior to setting up as well

* improve test

hussain-jafari committed:
pin pyarrow and add vscode dir to gitignore (#511)

* pin pyarrow and add vscode dir to gitignore

* Sbachmei/mic 6020/fix breaking test (#513)

* fix test_set_up_dask_clean_existing_cluster

* ignore mypy error

* check cluster prior to setting up as well

* improve test

* pin pyarrow and add vscode dir to gitignore

---------

Co-authored-by: Hussain Jafari <hjafari@uw.edu>
Co-authored-by: Steve Bachmeier <23350991+stevebachmeier@users.noreply.github.com>

stevebachmeier committed:
Sbachmei/merge main (#516)

hussain-jafari committed:
sequential row noising (#509)

Category: feature
JIRA issue: MIC-5982
Add tests for sequential row noising.
Read in data using load_standard_dataset and noise by looping through noise types, all shard-wise except RI with ACS, CPS, and WIC.

Testing
Ran new tests with ACS, CPS, and WIC on USA and RI and SSA on RI.

stevebachmeier committed:
Remove github workflow build.yml and update_readme.yml (#517)

hussain-jafari committed:
update census seed (#519)

Category: bugfix
JIRA issue: MIC-6045

Update census seed like we do with other datasets and add comments.

hussain-jafari committed:
create high noise configuration (#520)

Category: feature
JIRA issue: MIC-5865

Define a function which will return a custom config with a fixed value for all row noising values and for cell probabilities in column noising and default values otherwise.

Testing
Checked that function was returning a config that looked as expected for ACS and census.

hussain-jafari committed:
column noising (#521)

Category: feature
JIRA issue: MIC-5983

Add tests for column noising.
Update missingness in make_phonetic_errors.
Update get_common_datasets signature.
Move missingness test into column noising tests.

Testing
Ran tests on ACS, CPS, and WIC for RI and USA.

hussain-jafari committed:
swap month and day test correction (#522)

Category: feature
JIRA issue: MIC-6089

Exclude simulants whose date of birth have the same month and day from being considered for noising when calculating expected noise levels.

Testing
Ran on ACS, CPS, and WIC for USA.

hussain-jafari committed:
write placeholder tests (#527)

Category: feature
JIRA issue: MIC-6094

Add 3 tests which

pass for all parameter combinations
fail for parameter combinations with census and pass otherwise
runs for 5 minutes with tax_1040 and run quickly otherwise (the tax_1040 jobs will get cancelled when we request sruns for 1 minute)

Testing
Ran pytest command with acs, census, and tax_1040 and checked that all tests passed, the second test failed, and the third test ran for longer than 1 minute, respectively.

albrja committed:
Albrja/mic-6097/update-pandas-stubs-pin (#524)

Albrja/mic-6097/update-pandas-stubs-pin

Update pandas stubs pin
- *Category*: CI
- *JIRA issue*: [MIC-6097](https://jira.ihme.washington.edu/browse/MIC-6097)

<!-- 
Change description – why, what, anything unexplained by the above.
Include guidance to reviewers if changes are complex.
--> 

### Testing
<!--
Details on how code was verified, any unit tests local for the
repo, regression testing, etc. At a minimum, this should include an
integration test for a framework change. Consider: plots, images,
(small) csv file.

*** REMINDER ***
CI WILL NOT RUN ANY TESTS MARKED AS SLOW (CURRENTLY INCLUDES INTEGRATION TESTS).
MANUALLY RUN SLOW TESTS LIKE `pytest --runslow` WITH EACH PR.
-->
- [ ] all tests pass (`pytest --runslow`)

hussain-jafari committed:
post processing (#526)

Category: feature
JIRA issue: MIC-5984

Set up two post-processing tests (test dtypes and test that ID cols were unnoised) to work with sharded data.

Testing
Ran on all ACS, CPS, and WIC for RI.

hussain-jafari committed:
use copy to avoid updating data in place with dask (#529)

Category: bugfix
JIRA issue: MIC-6113

use copy to avoid updating data in place when using dask

Testing
Ran pytest--runslow.

hussain-jafari committed:
parallel script (#528)

Category: feature
JIRA issue: MIC-5513

Write script that allows to launch release tests in parallel based on parameters defined in csv.

Testing
Ran 3 tests where we expect 1) all tests to pass, 2) 2 tests to fail and the rest to pass, and 3) 2 tests to get cancelled and the rest to pass.

hussain-jafari committed:
special case duplicate_with_guardian (#532)

Category: feature
JIRA issue: MIC-6104

After applying duplicate_with_guardian noising and running its tests, perform subsequent noising on data before this noise was applied.

Testing
Ran census and checked that we weren't having errors from duplicate indexes.

hussain-jafari committed:
use LocalCluster in release tests (#531)

Category: feature
JIRA issue: MIC-6139

Testing
Ran release test on ACS RI data with dask.

hussain-jafari committed:
dask noising (#533)

Category: feature
JIRA issue: MIC-6139

Noise using dask.

Testing
Ran pytest with dask engine on subset of full USA_scale CPS and SSA data, and on full scale ACS data.

hussain-jafari committed:
define do_not_respond helper functions (#534)

Category: feature
JIRA issue: MIC-6152

Repurpose omit_row count function for use with do_not_respond and define fuzzy_checker function which accounts for oversampling in ACS and CPS.

Testing
Ran pytest on a few shards of ACS and CPS on USA data.

hussain-jafari committed:
fix NoiseTypes docstring noising order (#536)

fix NoiseTypes docstring noising order

hussain-jafari committed:
drop missing rows and check id cols (#535)

Category: feature
JIRA issue: MIC-6164

Drop missing rows after column noising and testing have been performed for a particular noise type.

Testing
Ran on subset of ACS and CPS USA data.
Put a breakpoint after dropping rows to check that they got dropped.

hussain-jafari committed:
guardian duplication tests with dask (#537)

Category: feature
JIRA issue: MIC-6153

Update dask tests to work with duplicate_with_guardian. This means:

defining how we get the info used for testing guardian duplication
testing guardian duplication using this info
updating fuzzy checking used for row noising to use kwargs

Testing
Ran pytest with dask on a few shards of USA data for ACS, CPS, and SSA (they don't perform guardian duplication but to test general row noising), and subsets of the shards of USA census.

hussain-jafari committed:
update script processing to handle new data (#539)

Category: feature
JIRA issue: MIC-6095
Process noise level and cpus per task.

Testing
Ran manual pytests and launched jobs through this script using a few shards and a subset of the rows for ACS, CPS, SSA, census, and tax_1040.

hussain-jafari committed:
don't fuzzy check empty shards (#540)

Category: bugfix
JIRA issue: MIC-6178

Don't perform fuzzy checks if no simulants were eligible for noising in a shard.
